### PR TITLE
Kernel+AudioServer: Use interrupts for Intel HDA audio buffer completion

### PR DESCRIPTION
**Pull request by gmta**
We used to not care about stopping an audio output stream for Intel HDA since AudioServer would continuously send new buffers to play. Since 707f5ac150ef858760eb9faa52b9ba80c50c4262 however, that has changed.

Intel HDA now uses interrupts to detect when each buffer was completed by the device, and uses a simple heuristic to detect whether a buffer underrun has occurred so it can stop the output stream.

This was tested on Qemu's Intel HDA (Linux x86_64) and a bare metal MSI Starship/Matisse HD Audio Controller.

CC @kleinesfilmroellchen

### CODE REVIEW

File: Kernel/Devices/Audio/IntelHDA/Stream.h [hunk 1]
```diff
@@ -6,26 +6,31 @@
...
     static constexpr u32 cyclic_buffer_size_in_ms = 40;
 
+    virtual ~Stream();
+
     u8 stream_number() const { return m_stream_number; }
     bool running() const { return m_running; }
     u32 sample_rate() const { return m_format_parameters.sample_rate; }
 
-    void start();
+    virtual void start();
```
Comment (kleinesfilmroellchen) [inline Kernel/Devices/Audio/IntelHDA/Stream.h:31]: Why does this need to be virtual? You don't override it in the OutputStream.

**Review by kleinesfilmroellchen — APPROVED** [review]

File: Kernel/Devices/Audio/IntelHDA/Stream.h (hunk 1 above)
Comment (gmta) [reply Kernel/Devices/Audio/IntelHDA/Stream.h:31]: It doesn't need to be, that slipped through :-) thanks!

File: Kernel/Devices/Audio/IntelHDA/Stream.h [hunk 2]
```diff
@@ -49,6 +54,24 @@ class Stream {
...
+        InterruptOnCompletionEnable = 1u << 2,
+    };
+
+    // 3.3.36 : Input/Output/Bidirectional Stream Descriptor Status
+    enum StreamStatusFlag : u8 {
+        BufferCompletionInterruptStatus = 1u << 2,
+    };
+
+    // 3.6.3: Buffer Descriptor List Entry
+    enum BufferDescriptorEntryFlag : u32 {
```
Comment (kleinesfilmroellchen) [inline Kernel/Devices/Audio/IntelHDA/Stream.h:66]: Those should both be enum classes as per standard coding style

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
Apart from the existing QEMU bugs, no regressions on WSL.

File: Kernel/Devices/Audio/IntelHDA/Stream.h (hunk 2 above)
Comment (gmta) [reply Kernel/Devices/Audio/IntelHDA/Stream.h:66]: Can you point me to the coding style where this is mentioned?

Comment (gmta) [reply Kernel/Devices/Audio/IntelHDA/Stream.h:66]: I know it's a best practice in general, but I don't think it's explicitly stated in our coding style. That being said, these new enums reflect the style in which all the other enums in the vicinity were coded and I'd rather not diverge from them or refactor them all in this PR. Additionally, the code would be littered with `to_underlying(..)` everywhere or I would need to have helper methods to keep the code readable - and that's what I was going for.

Comment (kleinesfilmroellchen) [reply Kernel/Devices/Audio/IntelHDA/Stream.h:66]: Okay, I guess it's fine in this context. For reference, @ldm5180 did a bunch of work converting enums to enum classes, and as a TL;DR even though to_underlying is necessary for many places, enum classes provide much-improved semantics over enums which are technically just C compatibility.

Comment (gmta) [reply Kernel/Devices/Audio/IntelHDA/Stream.h:66]: Yeah, a refactor is probably due for these enums, it's just not a lot of fun to do :^)

File: Kernel/Devices/Audio/IntelHDA/Controller.h [hunk 3]
```diff
@@ -26,7 +27,8 @@ class Codec;
 
 class Controller final
     : public AudioController
-    , public PCI::Device {
+    , public PCI::Device
+    , public IRQHandler {
```
Comment (supercomputer7) [inline Kernel/Devices/Audio/IntelHDA/Controller.h:31]: Hmm, is it possible that we could have a separate class to use `PCIIRQHandler` instead? does Intel HDA support MSI/MSI-X interrupts?

Comment (gmta) [reply Kernel/Devices/Audio/IntelHDA/Controller.h]: Looks like a drop-in replacement of `IRQHandler` by `PCIIRQHandler` worked just fine! Based on my tests HDA supports MSI-X.

**Pull request merged**